Add a migration notice on certain pages

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* Adds the migration and 'Not Now' notices
* On clicking 'Not Now', the migration notice never appears again for that user:
* Appears on Block Lab > Content Blocks, Block Lab > Settings, `/wp-admin/plugins.php`, and the Dashboard. 






#### Testing instructions
1. Go to Block Lab > Content Blocks, Block Lab > Settings, `/wp-admin/plugins.php`, and the Dashboard
2. Expected: the notice appears as shown above
3. Click 'Not Now'
4. Expected: the notice is hidden, and the 'Not Now' notice displays
5. Reload the page
6. Expected: The notice doesn't appear again
7. To make the notice appear again:
```sh
wp user list # Find your user ID from that
wp shell
update_user_meta( $user_id, 'block_lab_show_migration_notice', 1 ); // $user_id is the ID from the wp user list command.
```

kienstra committed:
Begin the migration notice

The notice still needs a lot of work,
but it now appears on the right pages.

kienstra committed:
Add a 'Learn more' link and fork a method to assert markup is equal

Thanks to Alain Schlesser for the method.
This simplifies the tests.

kienstra committed:
Enqueue the CSS and an empty script

The script will handle clicking 'Not now',
ensuring the user doesn't see it again,
and that a new notice displays.

kienstra committed:
On clicking 'Not now', hide the notice and display another

Also, add an AJAX call so that it
doesn't appear again once it's dismissed.

kienstra commented on js/admin.migration-notice.js:
```diff
@@ -0,0 +1,26 @@
+/* global XMLHttpRequest, FormData, ajaxurl */
+
+document.addEventListener( 'DOMContentLoaded', function() {
```
Haha, it's strange writing vanilla JS again

kienstra committed:
Fix unit tests on multisite

A user needs to be a superadmin
to be able to 'install_plugins'.

kienstra committed:
Remove helper function for equal markup, as it's no longer needed

It was a good function,
and might be useful later.

kienstra committed:
Improve the comments in the JS file

This is hard to read,
so these comments should help.

kienstra committed:
Replace '0' as the dismissed value

'dismissed' is a litte more semantic,
though 0 is used in Core,
like with 'show_welcome_panel'.

kienstra commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,205 @@
+					printf(
+						/* translators: %1$s: the plugin name */
+						esc_html__( 'The Block Lab team have moved. For future updates and improvements, migrate now to the new home of custom blocks: %1$s', 'block-lab' ),
+						'<strong>Genesis Custom Blocks.</strong>'
```
This isn't translated, as it's the name of the plugin. Maybe it should be 😄

kienstra commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,205 @@
+		return (
+			( isset( $screen->base, $screen->post_type ) && 'edit' === $screen->base && 'block_lab' === $screen->post_type )
+			||
+			( isset( $screen->base ) && in_array( $screen->base, [ 'plugins', 'dashboard', 'block_lab_page_block-lab-settings' ], true ) )
```
If you don't want this to display in the Dashboard, it'd be easy to remove `'dashboard'` here

kienstra committed:
Move the . after the <strong> tag

Also, improve some DocBlocks
and comments.

kienstra commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,207 @@
+					printf(
+						/* translators: %1$s: the plugin name */
+						esc_html__( 'The Block Lab team have moved. For future updates and improvements, migrate now to the new home of custom blocks: %1$s.', 'block-lab' ),
+						'<strong>Genesis Custom Blocks</strong>'
```
This isn't translated, as it's the name of the plugin. Maybe it should be 😄

lukecarbis replied:
I think it _should_ be translatable, but the translation should be "Genesis Custom Blocks" in every language.

So, yeah, switch to a `__()` here.

kienstra commented:
#### Request for review

Hi @lukecarbis,
Could you please review this when you can? The [top comment](https://github.com/getblocklab/block-lab/pull/635#issue-441126183) is up to date.

Also, the notice doesn't have borders between the the buttons, like in the Migration UX document. 

Are those borders intended, or are they just from the document formatting?

<img width="821" alt="borders-innotice" src="https://user-images.githubusercontent.com/4063887/86067831-67785e80-ba3b-11ea-8062-b02030a42935.png">

Thanks, Luke!

lukecarbis commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,207 @@
+				<p>
+					<?php
+					printf(
+						'<a target="_blank" href="https://getblocklab.com/docs/genesis-custom-blocks">%1$s</a>',
```
@RobStino This page doesn't exist yet. Do we have anything we can chuck up there, or do we at least have this page marked as a to do on your go to market list? If so, let's make sure the URL is correct.

RobStino replied:
No, this page doesn't exist yet, but I've got it marked as a to-do. I'll just need to make sure I use the correct URL.

lukecarbis commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,207 @@
+		check_ajax_referer( self::NOTICE_NONCE_ACTION, self::NOTICE_NONCE_NAME );
+
+		if ( ! current_user_can( self::NOTICE_CAPABILITY ) ) {
+			wp_die( -1 );
```
Why use `wp_die()` over [`wp_send_json_error()`](https://developer.wordpress.org/reference/functions/wp_send_json_error/)?

kienstra replied:
Good question, I'm fixing it 😄

kienstra replied:
Fixed in https://github.com/getblocklab/block-lab/pull/635/commits/d2ce757313994d0d487d55cbc9c17afe8d9f4e95

lukecarbis reviewed (changes requested):
Looking good Ryan! Couple of small questions, and minor change requested.

lukecarbis commented:
Tested, and everything works as expected.

I'll let @RobStino comment on the design side of things, I suspect there's a few tweaks he'd like to make.

kienstra committed:
Use wp_send_json_error(), translate plugin name

Luke had a good idea to use that instead of
wp_die( -1 ).
Also, the plugin name is now translated.

kienstra commented:
Hi @lukecarbis,
Thanks a lot for reviewing this, good points. They're applied above.

lukecarbis commented on php/admin/class-migration.php:
```diff
@@ -0,0 +1,210 @@
+
+		update_user_meta( get_current_user_id(), self::NOTICE_USER_META_KEY, self::NOTICE_DISMISSED_META_VALUE );
+
+		wp_die( 1 );
```
This should be `wp_send_json_success()`, I think.

kienstra replied:
Right, usually AJAX callbacks would use that.

This is a little different in that it doesn't send anything back to the [JS logic that called it](https://github.com/getblocklab/block-lab/pull/635/files#diff-682871fb6fb8d69d6b84890ae345f685R16).

lukecarbis replied:
I understand that we wouldn't be using any success response. It might be more efficient to just `wp_die()`.

Either way, we should try to be consistent. Use `wp_die()` for both the error and success cases (as it was prior to my `wp_send_json_error()` suggestion), or use `wp_send_json_success()` here. We don't have to provide any data, we can call that function without any parameters.

The only difference is that the unused output would be:
`{'success':true}`

Instead of:
`1`

kienstra replied:
Sounds good, how about just `wp_die()` for both?

kienstra replied:
Good point that it's still sending `1` or `-1` in the response.

kienstra replied:
https://github.com/getblocklab/block-lab/pull/635/commits/4e8167a4e178d46c4bd26852b73e6663183f1667 changes this to `wp_send_json_success()`

lukecarbis replied:
Thanks for 4e8167a – looks good. Would have been happy with `wp_die()`, but I like this approach better, personally, because it feels consistent with other AJAX responses.

kienstra replied:
Thanks, it is more like other AJAX calls 😄

lukecarbis reviewed (approved):
👍 from me. Just one minor suggestion, but no need for further review from me after that.

Would love @RobStino's signoff on this before it's merged, though.

kienstra committed:
Use wp_send_json_success() to be consistent

add some filteres to PHPUnit tests
for this

RobStino commented:
Hey Ryan. Thanks for the work on this.
If we can make some small adjustments to the design of the notice, that would be great. I'm keeping it super simple and inline with standard WP admin notifications.
Specific call outs:
- WP Admin Button styling to 2 primary actions
- Reduction of vertical spacing between text and "learn more" link
See below for my mock
![image](https://user-images.githubusercontent.com/1578148/86421653-2d9aa700-bd1e-11ea-8d0b-c555e5b55c77.png)

The second stage of the notification for when someone clicks "not now" is good as is.

kienstra commented:
Hi @RobStino,
Thanks, that looks good. I'll ping you here when it's applied.

kienstra committed:
Use traditional buttons for 'Not Now' and 'Migrate'

Using the normal button styling.
Also, add noopner noreferrer.

kienstra commented:
Hi @RobStino,
Thanks, I think your suggestion looks better.

How does it look with https://github.com/getblocklab/block-lab/pull/635/commits/0f1c6ea248101ac2b46e37ef2c995cb7af5b9573?

### Narrow screen

![narrow-width](https://user-images.githubusercontent.com/4063887/86487062-b287bf80-bd22-11ea-8dc3-bf7406138217.png)

### Wider screen

![wider-screen](https://user-images.githubusercontent.com/4063887/86487075-bca9be00-bd22-11ea-8af6-dbb1b45c766f.png)

There's no change in the 'Not now' notice:

![not-now-notice](https://user-images.githubusercontent.com/4063887/86487197-190cdd80-bd23-11ea-88e2-6e39f2754f7a.png)

Also, I [added](https://github.com/getblocklab/block-lab/pull/635/files#diff-24db437d8589eed1f645b12ea7e577a0R92) a `@todo` as a note to myself to check the link.

lukecarbis commented:
I think it needs a touch more padding to the left of the Not Now button.

RobStino commented:
Hey. Could you add 16px more margin to the left of the "Not Now" button?

kienstra commented:
Sure!

kienstra committed:
Add margin-left to the 'Not Now' button

This should give a little more space
between this and the copy to the left.

kienstra commented:
@RobStino,
Good idea. How does it look now with https://github.com/getblocklab/block-lab/pull/635/commits/4137f1addd19f0bb1f2277fdd8b05e910bdecc2d?

![notice-margin-left-full-screen](https://user-images.githubusercontent.com/4063887/86860122-e26bf580-c089-11ea-9499-956676c809aa.png)

RobStino reviewed (approved):
Looks good to me! Thanks for the work.

kienstra commented:
Nice, thanks for reviewing it!
